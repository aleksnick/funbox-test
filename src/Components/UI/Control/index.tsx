import React from "react";
import IWithStyles from "../../../Models/IWithStyles";
import Paper from "@material-ui/core/Paper";
import InputBase from "@material-ui/core/InputBase";
import Divider from "@material-ui/core/Divider";
import IconButton from "../IconButton";
import { ButtonColor } from "../Button";
import { withStyles } from "@material-ui/core";

const enterButtonKey = 13;

const styles = {
  root: {
    padding: "2px 4px",
    display: "flex",
    alignItems: "center",
    width: 400
  },
  input: {
    marginLeft: 8,
    flex: 1
  },
  iconButton: {
    padding: 10
  },
  divider: {
    width: 1,
    height: 28,
    margin: 4
  }
};

export interface ControlProps extends IWithStyles {
  placeholder: string;
  onSubmit?: (value: string) => void;
}

export interface ControlState {
  value: string;
}

/**
 * Инпут с кнопкой
 *
 * @export
 * @class Control
 * @extends {React.Component<ControlProps>}
 */
export class Control extends React.Component<ControlProps, ControlState> {
  constructor(props: ControlProps) {
    super(props);
    this.state = {
      value: ""
    };
  }

  render() {
    const { classes, placeholder, children } = this.props;
    return (
      <Paper className={classes["root"]} elevation={1}>
        <InputBase
          onChange={this.onChange}
          onKeyDown={this.onKeyPress}
          className={classes["input"]}
          placeholder={placeholder}
          value={this.state.value}
        />
        <Divider className={classes["divider"]} />
        <IconButton
          onClick={this.onSubmit}
          color={ButtonColor.Primary}
          className={classes["iconButton"]}
          aria-label="Directions"
        >
          {children}
        </IconButton>
      </Paper>
    );
  }

  onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.currentTarget.value;
    if (value === this.state.value) {
      return;
    }
    this.setState({
      value: value
    });
  };

  onKeyPress = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.keyCode === enterButtonKey) {
      this.onSubmit();
    }
  };

  onSubmit = () => {
    if (this.state.value === "") {
      return;
    }
    const onSubmit = this.props.onSubmit;
    if (onSubmit) {
      this.setState({
        value: ""
      });
      onSubmit(this.state.value.trim());
    }
  };
}

export default withStyles(styles)(Control);
