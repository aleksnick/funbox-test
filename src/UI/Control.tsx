import React from "react";
import IWithStyles from "../Models/IWithStyles";
import Paper from "@material-ui/core/Paper";
import InputBase from "@material-ui/core/InputBase";
import Divider from "@material-ui/core/Divider";
import IconButton from "./IconButton";
import { ButtonColor } from "./Button";
import { withStyles } from "@material-ui/core";

const styles = {
  root: {
    padding: "2px 4px",
    display: "flex",
    alignItems: "center",
    width: 400
  },
  input: {
    marginLeft: 8,
    flex: 1
  },
  iconButton: {
    padding: 10
  },
  divider: {
    width: 1,
    height: 28,
    margin: 4
  }
};

export interface ControlProps extends IWithStyles {
  placeholder: string;
  onSubmit?: (value: string) => void;
}

export interface ControlState {
  value: string;
}

/**
 * Badge
 *
 * @export
 * @class Control
 * @extends {React.Component<ControlProps>}
 */
export class Control extends React.Component<ControlProps, ControlState> {
  constructor(props: ControlProps) {
    super(props);
    this.state = {
      value: ""
    };
  }

  render() {
    const { classes, placeholder, children } = this.props;
    return (
      <Paper className={classes["root"]} elevation={1}>
        <InputBase onChange={this.onChange} className={classes["input"]} placeholder={placeholder} value={this.state.value} />
        <Divider className={classes["divider"]} />
        <IconButton
          onClick={this.onSubmit}
          color={ButtonColor.Primary}
          className={classes["iconButton"]}
          aria-label="Directions"
        >
          {children}
        </IconButton>
      </Paper>
    );
  }

  onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({
      value: e.currentTarget.value
    });
  };

  onSubmit = () => {
    const onSubmit = this.props.onSubmit;
    if (onSubmit) {
      this.setState({
        value: ""
      });
      onSubmit(this.state.value);
    }
  }
}

export default withStyles(styles)(Control);
