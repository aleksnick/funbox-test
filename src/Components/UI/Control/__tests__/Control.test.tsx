import React from "react";
import Control from "../";
import { render, shallow } from "enzyme";

describe("Control", () => {
  describe("Snapshots", () => {
    test("Simple Control snapshot", () => {
      const wrapper = render(<Control>Control</Control>);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("Check properties", () => {
    test("Control OnSubmit", () => {
      let changeValue = false;
      const onSubmit = () => {
        changeValue = true;
      };

      const wrapper = shallow(<Control onSubmit={onSubmit}>Control</Control>);
      wrapper.simulate("submit");

      expect(changeValue).toBe(true);
    });
  });
});
