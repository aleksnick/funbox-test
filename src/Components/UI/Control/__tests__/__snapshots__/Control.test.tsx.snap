// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Control Snapshots Simple Control snapshot 1`] = `
<div
  class="MuiPaper-root-5 MuiPaper-elevation1-8 MuiPaper-rounded-6 Control-root-1"
>
  <div
    class="MuiInputBase-root-32 Control-input-2"
  >
    <input
      class="MuiInputBase-input-42"
      type="text"
      value=""
    />
  </div>
  <hr
    class="MuiDivider-root-49 Control-divider-4"
  />
  <button
    class="MuiButtonBase-root-60 MuiIconButton-root-54 MuiIconButton-colorPrimary-56 Control-iconButton-3"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiIconButton-label-59"
    >
      Control
    </span>
  </button>
</div>
`;
